consensus/src/block_storage/execution_pool_test.rs

```diff
@@ -0,0 +1,734 @@
+
+    // Window size must be greater than 0, should panic
+    let window_size = 0;
+    block_store
+        .inner
+        .read()
+        .find_window_root(genesis_block.id(), window_size);
+}
+
+#[tokio::test]
+async fn test_window_root_no_forks() {
+    // window_size > NUM_BLOCKS
+    const NUM_BLOCKS: usize = 5;
+    let window_size: usize = 8;
+    let (_, block_store, pipelined_blocks) =
+        create_block_tree_no_forks::<{ NUM_BLOCKS }>(NUM_BLOCKS as u64, window_size).await;
+
+    // Genesis ──> A1 ──> ... ──> A4
+    let [genesis_block, a1, a2, _, a4] = pipelined_blocks;
+    let (commit_root, window_root) = get_roots(a4.block(), block_store.clone(), window_size);
+    let block_window =
+        get_blocks_from_block_store_and_window(block_store.clone(), a4.block(), window_size);
+
+    // commit_root
+    //      ↓
+    //              ┌──────────────────┐
+    //  Genesis ──> │ A1 ──> A2 ──> A3 │ ──> A4
+    //              └──────────────────┘
+    //      ↑                ↑
+    // window_root      block_window
+    // TODO WARNING should genesis ever be the window root??? Revisit
+    // NOTE: Window root calculations are done in two places: `find_window_root` and `find_root`, update
+    // both if needed
+    assert_eq!(commit_root.id(), genesis_block.id());
+    assert_eq!(block_window.first().unwrap().id(), a1.id());
+    assert_eq!(block_window.len(), 3);
+    assert_eq!(
+        window_root.expect("Window root not found"),
+        genesis_block.id()
+    );
+
+    // Prune A2
+    block_store.prune_tree(a2.id());
+    let (commit_root, window_root) = get_roots(a4.block(), block_store.clone(), window_size);
+    let block_window =
+        get_blocks_from_block_store_and_window(block_store.clone(), a4.block(), window_size);
+
+    //                   commit_root
+    //                        │
+    //              ┌──────── ↓ ───────┐
+    //  Genesis ──> │ A1 ──> A2 ──> A3 │ ──> A4
+    //              └──────────────────┘
+    //     ↑                  ↑
+    // window_root       block_window
+    assert_eq!(commit_root.id(), a2.id());
+    assert_eq!(block_window.first().unwrap().id(), a1.id());
+    assert_eq!(block_window.len(), 3);
+    assert_eq!(
+        window_root.expect("Window root not found"),
+        genesis_block.id()
+    );
+
+    // ----------------------------------------------------------------------------------------- //
+
+    // window_size < NUM_BLOCKS
+    let window_size: usize = 2;
+    let (_, block_store, pipelined_blocks) =
+        create_block_tree_no_forks::<{ NUM_BLOCKS }>(NUM_BLOCKS as u64, window_size).await;
+
+    // Genesis ──> A1 ──> ... ──> A4
+    let [genesis_block, _, a2, a3, a4] = pipelined_blocks;
+    let (commit_root, window_root) = get_roots(a4.block(), block_store.clone(), window_size);
+    let block_window =
+        get_blocks_from_block_store_and_window(block_store.clone(), a4.block(), window_size);
+
+    // commit_root     window_root
```

Comment:
commit_root should not be behind window_root, right?

Reply:
I think its confusing because we're not building the tree with things like `commit_callback`, which would move the `commit_root` up to point to the current block. This is why `commit_root` is behind `window_root` in the tests. But it should not happen in production. 

Also, for example, the behavior of `prune_tree` (which we use exclusively in tests and not in production code) is different than `commit_callback`. `commit_callback` would take in the  `current_block` as a param and move the `commit_root` to be that `current_block`. For contrast, in the tests I may call `prune_tree` and provide an older block (not the current block), which will move the `commit_root` to the block passed as a parameter to `prune_tree`, not necessarily the current block

Lmk if this makes sense. Put some diagrams below as well to illustrate the difference. 

<img width="1161" alt="Screenshot 2024-11-21 at 4 50 19 PM" src="https://github.com/user-attachments/assets/268626ab-6135-4475-9281-711abd1f80ad">
<img width="1051" alt="Screenshot 2024-11-21 at 4 56 44 PM" src="https://github.com/user-attachments/assets/03733e19-4f49-4a11-afa3-c9c55d360ba4">

Reply:
OK, I think the main confusion is that the window_root returned by get_roots is not the window_root of the block store, it's just "this window's root", while the commit_root is the actual commit_root of the block store.

I'm not sure how to make this less confusing; so I'm fine with it the way it is :P